src/java.base/unix/classes/sun/security/provider/NativePRNG.java

```diff
@@ -566,7 +579,31 @@ private void implNextBytes(byte[] data) {
                     }
                 } catch (IOException e){
                     throw new ProviderException("nextBytes() failed", e);
+                } finally {
+                    crLock.unlock();
                 }
         }
+
+        @Override
+        public void beforeCheckpoint(Context<? extends Resource> context) throws Exception {
+            crLock.lock();
+            mixRandom = null;
+            buffered = 0;
+            lastRead = 0;
+            for(int i=0; i<nextBuffer.length; i++) {
+                nextBuffer[i] = 0;
+            }
```

Comment:
For 2. I think it is just a matter of JDKResource.Priority. The use case you described is about dependent resources (see https://github.com/openjdk/crac/pull/8). In this case, the priority of the NativePRNG should be adjusted in the JDKResource.Priority.
Also, I would suggest adding a debug option that enables detection of the incorrect usage of the JDKResource A inside beforeCheckpoint of JDKResource B (priority A > priority B)

Reply:
Without the debug option, the image may be compromised -- so there should be no option of running without the checks. The currently employed ReentrantReadWriteLock does not block taking Read lock (access to PRNG) while holding Write lock (that is supposed to block access to the PRNG during checkpoint/restore -- but not from the same thread). It would be much more useful to throw an exception "the image is going to be wrong", or at worst to deadlock. But now the implementation does guarantee the security of the image.

https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/util/concurrent/locks/ReentrantReadWriteLock.html

Reply:
Thank you. Added the code to verify if NativePRNG is used from the beforeCheckpoint/afterRestore. In such a case, CheckpointException will be thrown.

Reply:
Thanks, now it looks good. Just as a note, in the similar case I've used IllegalSelectorException [1]. I think IllegalStateException may fit here as well -- we probably need to converge what exception(s) to use. But it is not necessary now.

[1] https://github.com/openjdk/crac/blob/43aba3d502832a5a3d2e9712558f62e0cf93dbbb/src/java.base/linux/classes/sun/nio/ch/EPollSelectorImpl.java#L384